base/broadcast.jl

```diff
@@ -864,13 +864,14 @@ broadcast_unalias(::Nothing, src) = src
 
 # Preprocessing a `Broadcasted` does two things:
 # * unaliases any arguments from `dest`
-# * "extrudes" the arguments where it is advantageous to pre-compute the broadcasted indices
-@inline preprocess(dest, bc::Broadcasted{Style}) where {Style} = Broadcasted{Style}(bc.f, preprocess_args(dest, bc.args), bc.axes)
-preprocess(dest, x) = extrude(broadcast_unalias(dest, x))
+# * calls `f` on the arguments (typically `extrude`, which pre-computes the broadcasted indices where advantageous)
+@inline preprocess(dest, bc) = preprocess(extrude, dest, bc)
+@inline preprocess(f, dest, bc::Broadcasted{Style}) where {Style} = Broadcasted{Style}(bc.f, preprocess_args(f, dest, bc.args), bc.axes)
+preprocess(f, dest, x) = f(broadcast_unalias(dest, x))
 
-@inline preprocess_args(dest, args::Tuple) = (preprocess(dest, args[1]), preprocess_args(dest, tail(args))...)
-preprocess_args(dest, args::Tuple{Any}) = (preprocess(dest, args[1]),)
-preprocess_args(dest, args::Tuple{}) = ()
+@inline preprocess_args(f, dest, args::Tuple) = (preprocess(f, dest, args[1]), preprocess_args(f, dest, tail(args))...)
+preprocess_args(f, dest, args::Tuple{Any}) = (preprocess(f, dest, args[1]),)
```

Comment:
Changing this to
```julia
Base.@propagate_inbounds preprocess_args(f, dest, args::Tuple) = (preprocess(f, dest, args[1]), preprocess_args(f, dest, tail(args))...)
Base.@propagate_inbounds preprocess_args(f, dest, args::Tuple{Any}) = (preprocess(f, dest, args[1]),)
```
get rid of the memory allocation in
```julia
using BenchmarkTools
u, k1, k2, k3, k4, k5, k6, k7 = (ones(1000) for i in 1:8)
function goo(u, k1, k2, k3, k4, k5, k6, k7)
  @. u = 0.1*(0.1*k1 + 0.2*k2 + 0.3*k3 + 0.4*k4 + 0.5*k5 + 0.6*k6 + 0.7*k7)
  nothing
end
@btime goo($u, $k1, $k2, $k3, $k4, $k5, $k6, $k7)
```
Before:
```julia
julia> @btime goo($u, $k1, $k2, $k3, $k4, $k5, $k6, $k7)
  707.291 ns (16 allocations: 416 bytes)
```
After:
```julia
julia> @btime goo($u, $k1, $k2, $k3, $k4, $k5, $k6, $k7)
  640.183 ns (0 allocations: 0 bytes)
```

Reply:
Thanks — the key is just `@inline`.